File: backend/app/schemas/custom_field.py
import uuid

from pydantic import BaseModel, ConfigDict

from app.models.enums import CustomFieldType


class CustomFieldBase(BaseModel):
    name: str
    field_type: CustomFieldType
    config: dict = {}


class CustomFieldCreate(CustomFieldBase):
    board_id: uuid.UUID


class CustomFieldRead(CustomFieldBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    board_id: uuid.UUID
    position: float


class CustomFieldValueUpsert(BaseModel):
    value: dict | None = None


class CustomFieldValueRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    custom_field_id: uuid.UUID
    card_id: uuid.UUID
    value: dict | None

File: frontend/src/types/customField.ts
export type CustomFieldType = "text" | "number" | "dropdown" | "checkbox" | "date";

export interface DropdownOption {
  id: string;
  label: string;
  color?: string;
}

export interface CustomField {
  id: string;
  name: string;
  field_type: CustomFieldType;
  config: { options?: DropdownOption[] } & Record<string, unknown>;
  board_id: string;
  position: number;
}

export interface CustomFieldCreate {
  name: string;
  field_type: CustomFieldType;
  config?: Record<string, unknown>;
  board_id: string;
}

export interface CustomFieldUpdate {
  name?: string;
  config?: Record<string, unknown>;
  position?: number;
}

export interface CustomFieldValue {
  id: string;
  custom_field_id: string;
  card_id: string;
  value: Record<string, unknown> | null;
}

export interface CustomFieldValueUpsert {
  value: Record<string, unknown> | null;
}
